### PR TITLE
refactor: replace O(n) array-shift send buffer with ring buffer

### LINKED ISSUE #27: perf: replace O(n) array-shift send buffer with ring buffer
Parent issue: wspulse/.github#28

## Problem

`client-ts` uses a plain array as its outbound send buffer. Dequeuing the front element with `Array.prototype.shift()` is O(n) because the runtime must copy all remaining elements forward. In a serial drain loop this produces O(n^2) total work for a full buffer flush.

## Proposed solution

Replace the array-based send buffer with a `RingBuffer<string | Uint8Array>` class using modular index arithmetic:

1. **New file `src/ring_buffer.ts`** -- `RingBuffer` class with:
   - Fixed capacity (set to `sendBufferSize`)
   - O(1) `push()` and O(1) `dequeue()` via head/tail pointers
   - Semantics matching `server/resume.go`

2. **Update `src/client.ts`** -- replace `(string | Uint8Array)[]` + `push()`/`shift()` with `RingBuffer` usage in the write loop

3. **Tests** -- unit tests for `RingBuffer` (capacity, wrap-around, full/empty edge cases) and updated component tests for send buffer behaviour

## Scope

- Internal refactor only -- no API changes, no behaviour changes
- `sendBufferSize` becomes the fixed ring buffer capacity

## Reference

- `wspulse/server/resume.go` -- existing ring buffer implementation for reference
- Parent issue wspulse/.github#28 for full context and cross-SDK comparison

### PR DESCRIPTION
**Pull request by maxence2997**
## Summary

Replace the array-based send buffer with a `RingBuffer` class using modular index arithmetic, achieving O(1) enqueue and O(1) dequeue. Previously `Array.prototype.shift()` was O(n) per dequeue, producing O(n^2) total work for a full buffer flush.

## Related issues

Closes #27
Relates to wspulse/.github#28

## Changes

- New internal `src/ring-buffer.ts`: generic `RingBuffer<T>` with fixed capacity, `push()`, `peek()`, `shift()`, `clear()`, all O(1) (clear is O(size) for GC slot release)
- `src/client.ts`: `sendBuffer` field changed from `(string | Uint8Array)[]` to `RingBuffer<string | Uint8Array>`; `send()` uses atomic `push()` return value; drain loop uses peek-then-shift to preserve frames on write failure; `shutdown()` uses `clear()`
- `CHANGELOG.md`: added Changed entry under Unreleased
- `test/ring-buffer.test.ts`: 15 unit tests covering construction, push, peek, shift, wrap-around, mid-cycle boundary enforcement, clear with non-zero head, and capacity-1 edge case

### Required

- [x] `make check` passes (format -> lint -> type-check -> test)
- [x] Each commit represents exactly one logical change
- [x] Commit messages follow the format in `commit-message-instructions.md`
- [x] No unrelated code reformatting in this PR

### CODE REVIEW

File: test/ring-buffer.test.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
+import { describe, it, expect } from "vitest";
+import { RingBuffer } from "../src/ring_buffer.js";
+
+describe("RingBuffer", () => {
+  // ── A: construction ────────────────────────────────────────────────
+
+  it("A1: starts empty with correct capacity", () => {
+    const rb = new RingBuffer<number>(4);
+    expect(rb.length).toBe(0);
```
[the remark this answers is not in the document]
Comment (maxence2997) [reply test/ring-buffer.test.ts:9]: Fixed in 91b52f5. Added push-boundary assertions to A1: fills to capacity (4 pushes succeed), confirms the 5th push returns false, and verifies length is 4.

File: src/client.ts [hunk 2]
```diff
@@ -594,7 +595,8 @@ class WspulseClient implements Client {
     try {
       while (this.sendBuffer.length > 0 && !this.closed) {
         if (this.ws.readyState !== WS_OPEN) return;
-        const encoded = this.sendBuffer[0];
+        const encoded = this.sendBuffer.peek();
+        if (encoded === undefined) break;
```
[the remark this answers is not in the document]
Comment (maxence2997) [reply src/client.ts]: Fixed in 451ff44. Changed break to shift()+continue so a stale undefined entry (unreachable under correct usage since T is now NonNullable) cannot cause repeated drain wakeups every 5ms.

File: src/ring_buffer.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+   * Append an item to the back of the buffer.
+   *
+   * @returns `true` if the item was added, `false` if the buffer is full.
+   */
+  push(item: T): boolean {
+    if (this.size >= this.cap) return false;
+    const index = (this.head + this.size) % this.cap;
+    this.data[index] = item;
+    this.size++;
+    return true;
```
[the remark this answers is not in the document]
Comment (maxence2997) [reply src/ring_buffer.ts]: Fixed in 451ff44. Constrained the class signature to RingBuffer<T extends NonNullable<unknown>>, preventing undefined from being pushed at compile time. The codec always returns string | Uint8Array so this is a zero-cost constraint for existing callers.

File: src/ring-buffer.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
+/**
+ * Fixed-capacity circular buffer with O(1) push and O(1) shift.
+ *
+ * Used internally as the outbound send buffer. Not exported from the
+ * package — consumers interact with `sendBufferSize` via options.
+ *
+ * @internal
+ */
```
[the remark this answers is not in the document]
Comment (maxence2997) [reply src/ring-buffer.ts:8]: Fixed in 21b9bfe. Renamed `src/ring_buffer.ts` to `src/ring-buffer.ts` to match the test file `ring-buffer.test.ts` (both now use hyphen). Updated the two affected import paths.

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,10 @@
 
 ## [Unreleased]
 
+### Changed
+
+- Internal send buffer replaced with a ring buffer for O(1) enqueue and dequeue (previously O(n) dequeue via `Array.shift()`). No API or behaviour changes.
+
```
[the remark this answers is not in the document]
Comment (maxence2997) [reply CHANGELOG.md:8]: Fixed — updated PR description to `src/ring-buffer.ts` (hyphen) to match the renamed file.

File: src/ring-buffer.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+export class RingBuffer<T extends NonNullable<unknown>> {
+  private readonly data: (T | undefined)[];
+  private head = 0;
+  private size = 0;
+  private readonly cap: number;
+
+  constructor(capacity: number) {
```
[the remark this answers is not in the document]
Comment (maxence2997) [reply src/ring-buffer.ts:15]: Won't fix. `RingBuffer` is `@internal` and is only instantiated in one place — `client.ts` constructor — using `opts.sendBufferSize`, which is already validated upstream in `validateOptions()` (`Number.isFinite` + `Number.isInteger` + range `[1, 4096]`). Adding duplicate validation in an internal class adds code for a scenario that cannot occur through the public API.
